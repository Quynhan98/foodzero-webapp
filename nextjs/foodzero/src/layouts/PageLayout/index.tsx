import React, { memo, useEffect, useState } from 'react'
import { useRouter } from 'next/router'
import { Box, Center } from '@chakra-ui/react'

// Components
import Header from '@layouts/Header'
import Footer from '@layouts/Footer'

// Constants
import { LOCAL_STORAGE_KEY, PAGE_URL } from '@constants/index'

// Context
import { AuthProvider } from '@contexts/AuthProvider'

// Services
import { getLocalStorage } from '@utils/localStorage'

// Hooks
import { useAuthContext } from '@hooks/useAuthContext'

export interface IPageLayoutsProps {
  children: React.ReactNode
}

const PageLayouts = ({ children }: IPageLayoutsProps) => {
  const router = useRouter()
  const { pathname } = router
  const { userId } = useAuthContext()
  const [authorized, setAuthorized] = useState(false)

  const isLoginPage = pathname === PAGE_URL.LOGIN.URL
  const hasAccount = !!getLocalStorage(LOCAL_STORAGE_KEY.IS_TOKEN)

  const authCheck = (isLogin: boolean) => {
    // Redirect to Login page with unauthorized user
    if (isLogin || !!userId) {
      setAuthorized(true)
    } else {
      setAuthorized(false)

      router.push(PAGE_URL.LOGIN.URL)
    }
  }

  useEffect(() => {
    authCheck(hasAccount)

    const hideContent = () => setAuthorized(false)

    // Hide page content on route change
    if (!hasAccount) router.events.on('routeChangeStart', hideContent)

    router.events.on('routeChangeComplete', authCheck)

    return () => {
      router.events.off('routeChangeStart', hideContent)
      router.events.off('routeChangeComplete', authCheck)
    }
  }, [hasAccount, router.events, router.route, userId])

  if (!authorized) return null

  return (
    <AuthProvider>
      {isLoginPage ? (
        <Center
          as="main"
          minHeight="100vh"
          pb="150px"
          backgroundColor="alabaster"
        >
          {children}
        </Center>
      ) : (
        <>
          <Header />
          <Box
            as="main"
            minHeight="100vh"
            maxWidth="1664px"
            width="100%"
            margin="0 auto"
            padding={{ base: '12px', lg: '0px' }}
          >
            {children}
          </Box>
          <Footer />
        </>
      )}
    </AuthProvider>
  )
}

export default memo(PageLayouts)
