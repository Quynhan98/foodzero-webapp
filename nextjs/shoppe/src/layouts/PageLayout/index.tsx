import Head from 'next/head'
import { Box } from '@chakra-ui/react'
import { ReactNode } from 'react'

// Layouts
import Header from '@layouts/Header'
import Footer from '@layouts/Footer'

// Hooks
import { useCartContext } from '@hooks/useCartContext'
import { useLoadingContext } from '@hooks/useLoadingContext'

// Components
import LoadingIndicator from '@components/LoadingIndicator'

interface PageLayoutProps {
  children: ReactNode
}

const PageLayout = ({ children }: PageLayoutProps) => {
  const { listCart } = useCartContext()
  const { loading } = useLoadingContext()

  return (
    <>
      <Head>
        <title>Shoppe</title>
        <meta name="title" content="Shoppe" />
        <link rel="icon" href="/images/logo.png" />
      </Head>
      <Header carts={listCart} />
      <Box
        as="main"
        minHeight="100vh"
        maxWidth="1248px"
        width="100%"
        margin="0 auto"
        padding={{ base: '16px', lg: '0px' }}
      >
        {children}
      </Box>
      <Footer />
      {loading && <LoadingIndicator size="lg" />}
    </>
  )
}

export default PageLayout
