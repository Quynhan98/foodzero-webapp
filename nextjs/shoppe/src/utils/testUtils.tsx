import { ReactNode } from 'react'
import { render } from '@testing-library/react'
import { ChakraProvider } from '@chakra-ui/react'

// Themes
import { customTheme } from '@themes/index'

export const customRender = (component: ReactNode) => {
  return render(
    <ChakraProvider theme={customTheme}>{component}</ChakraProvider>,
  )
}

export * from '@testing-library/react'
export { customRender as render }
