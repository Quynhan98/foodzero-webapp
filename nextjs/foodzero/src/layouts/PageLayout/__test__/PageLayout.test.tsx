// Utils
import { render } from '@utils/testUtils'

// Layouts
import PageLayout from '@layouts/PageLayout'

// Utils
import { NextRouterProvider } from '@utils/nextRouterProvider'

// Constants
import { PAGE_URL } from '@constants/index'

describe('PageLayout render', () => {
  beforeEach(() => {
    window.localStorage.clear()
  })

  it('Should PageLayout match Snapshot', () => {
    window.localStorage.setItem('userId', JSON.stringify('gff1'))
    window.localStorage.setItem('isToken', JSON.stringify('gff1-LP'))

    const { container } = render(
      <NextRouterProvider router={{ pathname: PAGE_URL.HOME.URL }}>
        <PageLayout>Contents Page</PageLayout>
      </NextRouterProvider>,
    )

    expect(container).toMatchSnapshot()
  })

  it('Should PageLayout of Login page match Snapshot', () => {
    const { container } = render(
      <NextRouterProvider router={{ pathname: PAGE_URL.LOGIN.URL }}>
        <PageLayout>Contents Login Page</PageLayout>
      </NextRouterProvider>,
    )

    expect(container).toMatchSnapshot()
  })
})
