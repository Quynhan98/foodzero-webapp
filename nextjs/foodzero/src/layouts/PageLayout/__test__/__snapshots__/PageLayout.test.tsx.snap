// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageLayout render Should PageLayout match Snapshot 1`] = `
<div>
  <header
    class="chakra-container css-1qxfw8o"
  >
    <div
      class="css-f1oc3g"
    >
      <h1
        class="chakra-heading css-16unufg"
      >
        <a
          class="chakra-link css-spn4bz"
          href="/"
        >
          <div
            class="css-1pthsoj"
          >
            <img
              alt="Logo FoodZero"
              data-nimg="fill"
              decoding="async"
              sizes="(max-width: 768px) 100vw,
 (max-width: 1200px) 50vw,
 33vw"
              src="/_next/image?url=%2Fimages%2Flogo.webp&w=3840&q=75"
              srcset="/_next/image?url=%2Fimages%2Flogo.webp&w=256&q=75 256w, /_next/image?url=%2Fimages%2Flogo.webp&w=384&q=75 384w, /_next/image?url=%2Fimages%2Flogo.webp&w=640&q=75 640w, /_next/image?url=%2Fimages%2Flogo.webp&w=750&q=75 750w, /_next/image?url=%2Fimages%2Flogo.webp&w=828&q=75 828w, /_next/image?url=%2Fimages%2Flogo.webp&w=1080&q=75 1080w, /_next/image?url=%2Fimages%2Flogo.webp&w=1200&q=75 1200w, /_next/image?url=%2Fimages%2Flogo.webp&w=1920&q=75 1920w, /_next/image?url=%2Fimages%2Flogo.webp&w=2048&q=75 2048w, /_next/image?url=%2Fimages%2Flogo.webp&w=3840&q=75 3840w"
              style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: cover; color: transparent;"
            />
          </div>
        </a>
      </h1>
      <button
        aria-label="menu"
        class="chakra-button css-1trbkkh"
        data-testid="buttonMenu"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-1k6ehyi"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M 3 5 A 1.0001 1.0001 0 1 0 3 7 L 21 7 A 1.0001 1.0001 0 1 0 21 5 L 3 5 z M 3 11 A 1.0001 1.0001 0 1 0 3 13 L 21 13 A 1.0001 1.0001 0 1 0 21 11 L 3 11 z M 3 17 A 1.0001 1.0001 0 1 0 3 19 L 21 19 A 1.0001 1.0001 0 1 0 21 17 L 3 17 z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
    <div
      class="css-1y4a8vd"
    >
      <a
        class="chakra-link css-spn4bz"
        href="tel:+86 852 346 000"
      >
        <p
          class="chakra-text css-z99krb"
        >
          +86 852 346 000
        </p>
      </a>
      <button
        class="chakra-button css-555q6y"
        type="button"
      >
        Reservations
      </button>
    </div>
  </header>
  <main
    class="css-1liisy1"
  >
    Contents Page
  </main>
  <footer
    class="css-1d0otq1"
  >
    <div
      class="css-15ms0zq"
    >
      <div
        class="css-cl2fjb"
      >
        <h1
          class="chakra-heading css-gm12et"
        >
          Food Zero.
        </h1>
        <div
          class="css-0"
        >
          <p
            class="chakra-text css-ppz0ah"
          >
            Contact
          </p>
          <a
            href="tel:+86 852 346 000"
          >
            <p
              class="chakra-text css-gz08eg"
            >
              +86 852 346 000
            </p>
          </a>
          <a
            href="mailto:info@foodzero.com"
          >
            <p
              class="chakra-text css-1oecnb9"
            >
              info@foodzero.com
            </p>
          </a>
          <p
            class="chakra-text css-15kw7zj"
          >
            1959 Sepulveda Blvd. Culver City, CA, 90230
          </p>
        </div>
      </div>
      <div
        class="css-0"
      >
        <p
          class="chakra-text css-1g7063y"
        >
          Never Miss a Recipe
        </p>
        <div
          class="css-1jn9ke7"
        >
          <input
            class="chakra-input css-1rjerp5"
            placeholder="Email Address"
          />
          <button
            class="chakra-button css-1a1bbvk"
            type="button"
          >
            Subscribe
          </button>
        </div>
        <p
          class="chakra-text css-3firl6"
        >
          Join our subscribers and get best recipe delivered each week!
        </p>
      </div>
    </div>
    <hr
      aria-orientation="horizontal"
      class="chakra-divider css-dua0ms"
    />
    <div
      class="css-1b9jvdy"
    >
      <p
        class="chakra-text css-8wvo7l"
      >
        Â© 2020 Zero Inc. All rights Reserved
      </p>
      <ul
        class="css-shidlx"
        role="list"
      >
        <li
          class="css-155za0w"
        >
          <a
            href="/"
          >
            <img
              alt="Instagram"
              data-nimg="1"
              decoding="async"
              height="28"
              loading="lazy"
              src="/_next/image?url=%2Fimages%2Finstagram.webp&w=64&q=75"
              srcset="/_next/image?url=%2Fimages%2Finstagram.webp&w=32&q=75 1x, /_next/image?url=%2Fimages%2Finstagram.webp&w=64&q=75 2x"
              style="color: transparent;"
              value="1"
              width="28"
            />
          </a>
        </li>
        <li
          class="css-155za0w"
        >
          <a
            href="/"
          >
            <img
              alt="Twitter"
              data-nimg="1"
              decoding="async"
              height="28"
              loading="lazy"
              src="/_next/image?url=%2Fimages%2Ftwitter.webp&w=64&q=75"
              srcset="/_next/image?url=%2Fimages%2Ftwitter.webp&w=32&q=75 1x, /_next/image?url=%2Fimages%2Ftwitter.webp&w=64&q=75 2x"
              style="color: transparent;"
              value="2"
              width="28"
            />
          </a>
        </li>
        <li
          class="css-155za0w"
        >
          <a
            href="/"
          >
            <img
              alt="Facebook"
              data-nimg="1"
              decoding="async"
              height="28"
              loading="lazy"
              src="/_next/image?url=%2Fimages%2Ffacebook.webp&w=64&q=75"
              srcset="/_next/image?url=%2Fimages%2Ffacebook.webp&w=32&q=75 1x, /_next/image?url=%2Fimages%2Ffacebook.webp&w=64&q=75 2x"
              style="color: transparent;"
              value="3"
              width="28"
            />
          </a>
        </li>
        <li
          class="css-155za0w"
        >
          <a
            href="/"
          >
            <img
              alt="youtube"
              data-nimg="1"
              decoding="async"
              height="28"
              loading="lazy"
              src="/_next/image?url=%2Fimages%2Fyoutube.webp&w=64&q=75"
              srcset="/_next/image?url=%2Fimages%2Fyoutube.webp&w=32&q=75 1x, /_next/image?url=%2Fimages%2Fyoutube.webp&w=64&q=75 2x"
              style="color: transparent;"
              value="4"
              width="28"
            />
          </a>
        </li>
      </ul>
    </div>
  </footer>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;

exports[`PageLayout render Should PageLayout of Login page match Snapshot 1`] = `
<div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
